pkg/lib/import-json.ts

```diff
@@ -0,0 +1,182 @@
+   will catch the ValidationError and return a fallback value.
+
+   Validation is strict: If a validation error occurs deep inside a
+   nested structure, the whole structure is rejected.
+ */
+
+/* WRITING IMPORTER FUNCTIONS
+
+   The process of writing a importer function for a given TypeScript
+   interface is pretty mechanic, and could well be automated.
+
+   For example, here are the functions for Player and Team interfaces:
+
+       interface Player {
+          name: string;
+          age: number | undefined;
+       }
+
+       function import_Player(val: JsonValue): Player {
+           const obj = import_json_object(val);
+           return {
+               name: import_mandatory(obj, "name", import_string),
+               age: import_optional(obj, "age", import_number),
+           };
+       }
+
+       interface Team {
+          name: string;
+          players: Player[];
+       }
+
+       function import_Team(val: JsonValue): Team {
+           const obj = import_json_object(val);
+           return {
+               name: import_mandatory(obj, "name", import_string),
+               players: import_mandatory(obj, "players", v => import_array(v, import_Player)),
+           }
+       }
+
+   This way, TypeScript will check that the returned values are indeed
+   valid for their type declaration. You can't get that wrong. What is
+   not checked is that you use the right field names when accessing
+   input JsonObjects. But we could write a linter function for that.
+
+   More examples can be found in "pkg/shell/manifests.ts".
+ */
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string): never {
+    throw new ValidationError(`JSON validation error for ${validation_path.join("")}: ${msg}`);
+}
+
+export function import_string(val: JsonValue): string {
+    if (typeof val == "string")
+        return val;
+    validation_error(`Not a string: ${JSON.stringify(val)}`);
+}
+
+export function import_number(val: JsonValue): number {
+    if (typeof val == "number")
+        return val;
+    validation_error(`Not a number: ${JSON.stringify(val)}`);
+}
+
+export function import_boolean(val: JsonValue): boolean {
+    if (typeof val == "boolean")
+        return val;
+    validation_error(`Not a boolean: ${JSON.stringify(val)}`);
+}
+
+function is_json_object(val: JsonValue): val is JsonObject {
+    return !!val && typeof val == "object" && !Array.isArray(val);
+}
+
+export function import_json_object(val: JsonValue): JsonObject {
+    if (is_json_object(val))
+        return val;
+    validation_error(`Not an object: ${JSON.stringify(val)}`);
+}
+
+function is_json_array(val: JsonValue): val is JsonValue[] {
+    return !!val && typeof val == "object" && Array.isArray(val);
```

Comment:
`isArray()` already deals fine with `undefined`, `null`, `NaN`, `{}` or other garbage. So the only reason to have your own function would be if you inspect the *contents* of `val`, i.e. that it doesn't contain some entries that are not representable in  JSON. But being a `JsonValue` ought to already enforce that, so I think this function is unnecessary.

Reply:
Good point! Fixed.